File: paleo_workbench/ui/pages/seismic_prediction_page.py
from __future__ import annotations

from PySide6.QtCore import Signal, Slot
from PySide6.QtGui import QCloseEvent
from PySide6.QtWidgets import QHBoxLayout, QMessageBox, QVBoxLayout, QWidget

from paleo_workbench.ui import tokens
from paleo_workbench.ui.owned_worker_job import OwnedWorkerJob
from paleo_workbench.viz.prediction_helpers import active_prediction_task
from paleo_workbench.ui.pages.seismic_attribute_panel import SeismicAttributePanel
from paleo_workbench.ui.pages.seismic_control_panel import SeismicControlPanel
from paleo_workbench.ui.pages.seismic_context_toolbar import SeismicContextToolbar
from paleo_workbench.ui.pages.seismic_view_panel import SeismicViewPanel
from paleo_workbench.workflow.stratigraphy import active_target_horizon
from paleo_workbench.catalog import get_catalog_service
from paleo_workbench.prediction.inference_service import (
    link_run_to_domain_task,
    materialize_prediction_task,
    resolve_prediction_inputs,
    start_inference,
)
from paleo_workbench.prediction.inference_worker import InferenceWorker
from paleo_workbench.prediction.providers import (
    CAPABILITY_FACIES,
    MODEL_ID_DEMO,
    ensure_default_models,
)


class SeismicPredictionPage(QWidget):
    """Reference-style seismic analysis workbench around the existing view."""

    prediction_updated = Signal()
    send_to_mapping_requested = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("SeismicPredictionPage")
        self._project = None
        self._tasks: list = []
        self._inference_service = None
        self._inference_job = OwnedWorkerJob(self)
        self._session_token = object()
        self._active_inference_context = None

        outer = QVBoxLayout(self)
        outer.setContentsMargins(
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
            tokens.PAGE_MARGIN,
        )
        outer.setSpacing(tokens.SPACE_4)

        self.context_toolbar = SeismicContextToolbar()
        outer.addWidget(self.context_toolbar)

        content = QHBoxLayout()
        content.setSpacing(tokens.SPACE_4)

        self.attribute_panel = SeismicAttributePanel()
        content.addWidget(self.attribute_panel, 0)

        self.view_panel = SeismicViewPanel()
        content.addWidget(self.view_panel, 1)

        self.control_panel = SeismicControlPanel()
        content.addWidget(self.control_panel, 0)

        outer.addLayout(content, 1)

        self.context_toolbar.run_requested.connect(self._on_run)
        self.context_toolbar.demo_requested.connect(self._on_demo)
        self.control_panel.send_requested.connect(self.send_to_mapping_requested.emit)
        self.control_panel.display_mode_changed.connect(self.view_panel.set_display_mode)
        self.attribute_panel.attribute_changed.connect(self._on_attribute)
        self.control_panel.well_tie_toggled.connect(self.view_panel.set_well_tie_enabled)
        self.view_panel.view_ready.connect(self._on_view_ready)

    def set_project(self, project) -> None:
        if project is not self._project:
            self._session_token = object()
        self._project = project

    def shutdown_workers(self, wait_ms: int = 3_000) -> bool:
        """Stop project-owned inference before its catalog is rebound."""
        joined = self._inference_job.shutdown(wait_ms)
        # A switch is aborted when a job cannot join.  Preserve the current
        # page/session in that case; result signals were disconnected by the
        # owned job and the old catalog remains open for the detached worker.
        if not joined:
            return False
        self._session_token = object()
        self._active_inference_context = None
        self._inference_service = None
        self._project = None
        shutdown = getattr(self.view_panel, "shutdown", None)
        if callable(shutdown):
            shutdown()
        return joined

    def closeEvent(self, event: QCloseEvent) -> None:  # noqa: N802
        # A page owns its inference thread; joining here covers direct widget
        # destruction as well as normal project-session shutdown.
        self.shutdown_workers()
        event.accept()

    def update_state(self, prediction_tasks: list | tuple | None, project=None) -> None:
        if project is not None:
            self.set_project(project)
        self._tasks = list(prediction_tasks or [])
        task = self._current_task()
        self.view_panel.update_state(task, project=self._project)
        self.control_panel.update_state(task, self.view_panel.volume_shape)
        self._sync_workbench_context(task)
        self.control_panel.set_controls_enabled(self.view_panel.is_view_ready())

    def _current_task(self):
        return active_prediction_task(self._tasks)

    def _on_attribute(self, label: str) -> None:
        self.view_panel.set_attribute_label(label)
        self.attribute_panel.set_selected_attribute(label)
        self._sync_workbench_context(self._current_task())

    def _on_view_ready(self, enabled: bool) -> None:
        self.control_panel.set_controls_enabled(enabled)
        if enabled:
            self._sync_workbench_context(self._current_task())

    def _sync_workbench_context(self, task) -> None:
        attribute = self.view_panel.attribute_label()
        mode = self.view_panel.display_mode()
        self.attribute_panel.set_selected_attribute(attribute)
        self.control_panel.set_attribute_label(attribute)
        self.context_toolbar.set_context(
            task,
            self.control_panel.horizon_value.text(),
            attribute,
            mode,
        )

    def _on_run(self) -> None:
        """Production inference: resolve a production model, never auto-run mock.

        No production model in the ModelRegistry → explicit 「未配置生产模型」
        unavailable state (spec P2 §3d). Demo requires explicit demo mode.
        """
        if self._project is None:
            QMessageBox.warning(self, "地震预测", "未绑定工程，无法运行")
            return
        service = get_catalog_service()
        if service is None:
            QMessageBox.warning(self, "地震预测", "未连接数据目录，无法运行推断")
            return
        ensure_default_models(service)
        model_version = service.find_production_model(CAPABILITY_FACIES)
        if model_version is None:
            QMessageBox.warning(
                self,
                "地震预测",
                "未配置生产模型，无法运行科学预测。\n"
                "请先注册生产模型（ModelRegistry），或通过「运行演示预测」查看演示结果。",
            )
            return
        self._start_inference(
            service,
            model_version.id,
            workflow="seismic_facies",
            name_prefix="地震相预测",
            demo=False,
        )

    def _on_demo(self) -> None:
        """Explicit demo mode: run the registered DemoModelProvider."""
        if self._project is None:
            QMessageBox.warning(self, "地震预测", "未绑定工程，无法运行")
            return
        service = get_catalog_service()
        if service is None:
            QMessageBox.warning(self, "地震预测", "未连接数据目录，无法运行推断")
            return
        ensure_default_models(service)
        try:
            demo_version = service.get_model_version(MODEL_ID_DEMO, "1")
        except Exception as exc:
            QMessageBox.warning(self, "地震预测", f"演示模型未注册: {exc}")
            return
        self._start_inference(
            service,
            demo_version.id,
            workflow="seismic_facies",
            name_prefix="地震相预测(Demo)",
            demo=True,
        )

    def _start_inference(
        self,
        service,
        model_version_id: str,
        *,
        workflow: str,
        name_prefix: str,
        demo: bool,
    ) -> None:
        if self._inference_job.is_running:
            return
        self._inference_service = service
        if demo:
            input_ids = []
        else:
            from paleo_workbench.prediction.inference_service import (
                resolve_inputs_for_model,
            )
            from paleo_workbench.prediction.input_contract import InputContractError

            try:
                input_ids = resolve_inputs_for_model(
                    self._project, service, model_version_id, strict=True
                )
            except InputContractError as exc:
                QMessageBox.warning(self, "地震预测", f"输入不满足模型契约: {exc}")
                return
        run = start_inference(
            service,
            model_version_id=model_version_id,
            input_version_ids=input_ids,
            parameters={
                "seed": len(self._tasks),
                "workflow": workflow,
                "name_prefix": name_prefix,
                "demo": demo,
            },
        )
        worker = InferenceWorker(service, run.id)
        self._active_inference_context = (self._session_token, self._project, service)
        self._inference_job.start(
            worker,
            terminal_signals=(worker.terminal,),
            result_connections=(
                (worker.completed, self._on_inference_completed_if_current),
                (worker.failed, self._on_inference_failed_if_current),
            ),
            target=name_prefix,
        )

    @Slot(object)
    def _on_inference_completed_if_current(self, payload) -> None:
        context = self._active_inference_context
        if context != (self._session_token, self._project, self._inference_service):
            return
        self._on_inference_completed(payload)

    @Slot(str)
    def _on_inference_failed_if_current(self, text: str) -> None:
        context = self._active_inference_context
        if context is None or context[:2] != (self._session_token, self._project):
            return
        self._on_inference_failed(text)

    def _on_inference_completed(self, payload: dict) -> None:
        run = payload.get("run")
        if run is None or getattr(run, "status", "") == "failed":
            error = "未知错误"
            if run is not None:
                error = (run.parameters or {}).get("error", "未知错误")
            QMessageBox.warning(self, "地震预测失败", f"推断失败: {error}")
            return
        result = payload.get("result")
        if not result:
            error = (getattr(run, "parameters", None) or {}).get("error") or (
                "预测完成但未返回可用结果"
            )
            QMessageBox.warning(self, "地震预测失败", f"推断失败: {error}")
            return
        params = result.get("parameters") or {}
        factor_ids = [
            task.id
            for task in self._project.factor_map_tasks
            if getattr(task, "status", "") == "complete"
        ]
        out_vids = list(getattr(run, "output_version_ids", None) or [])
        task = materialize_prediction_task(
            self._project,
            result,
            name_prefix=params.get("name_prefix", "地震相预测"),
            workflow=params.get("workflow", "seismic_facies"),
            target_horizon=(
                active_target_horizon(self._project)
                or self._project.stratigraphy.target_horizon
                or ""
            ),
            factor_map_ids=factor_ids,
            run_id=str(getattr(run, "id", "") or ""),
            output_version_id=str(out_vids[0]) if out_vids else "",
        )
        self._project.prediction_tasks.append(task)
        if self._inference_service is not None:
            try:
                link_run_to_domain_task(self._inference_service, run.id, task.id)
            except Exception:
                # Provenance-critical bridge: record the failure instead of
                # silently dropping it (H3 — downstream lineage would be empty).
                import logging

                logging.getLogger(__name__).warning(
                    "link_run_to_domain_task failed for run %s task %s",
                    run.id,
                    task.id,
                    exc_info=True,
                )
                task.model_metadata = dict(task.model_metadata or {})
                task.model_metadata["link_failed"] = True
        self._tasks = list(self._project.prediction_tasks)
        self.update_state(self._tasks, project=self._project)
        self.prediction_updated.emit()

    def _on_inference_failed(self, text: str) -> None:
        QMessageBox.critical(self, "地震预测失败", f"推断失败: {text}")

File: libs/ui_wellseis/src/qt/seismic_prediction_page.cpp
#include <pwb/ui_wellseis/qt/seismic_prediction_page.hpp>

#include <QCloseEvent>
#include <QComboBox>
#include <QMessageBox>
#include <QSplitter>
#include <QVBoxLayout>

#include <pwb/ui_wellseis/json_helpers.hpp>
#include <pwb/ui_wellseis/output_labels.hpp>
#include <pwb/ui_wellseis/qt/seismic_attribute_panel.hpp>
#include <pwb/ui_wellseis/qt/seismic_context_toolbar.hpp>
#include <pwb/ui_wellseis/qt/seismic_control_panel.hpp>
#include <pwb/ui_wellseis/qt/seismic_view_panel.hpp>
#include <pwb/ui_wellseis/resource_sources.hpp>
#include <pwb/ui_wellseis/task_state.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

constexpr int kPanelMaxWidth = 360;

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

std::vector<ResourceSlice> segy_resources(const ProjectSlice* project) {
    std::vector<ResourceSlice> out;
    if (project == nullptr) {
        return out;
    }
    for (const auto& resource : project->resources) {
        if (is_segy_resource(resource)) {
            out.push_back(resource);
        }
    }
    return out;
}

}  // namespace

SeismicPredictionPage::SeismicPredictionPage(QWidget* parent)
    : QWidget(parent) {
    setObjectName(QStringLiteral("SeismicPredictionPage"));

    auto* outer = new QVBoxLayout(this);
    outer->setContentsMargins(16, 16, 16, 16);
    outer->setSpacing(12);

    context_toolbar_ = new SeismicContextToolbar(this);
    outer->addWidget(context_toolbar_);

    splitter_ = new QSplitter(Qt::Horizontal, this);
    splitter_->setObjectName(QStringLiteral("SeismicPredictionSplitter"));

    // 属性 | 视图 | 控制: the view stays the stretchy center.
    attribute_panel_ = new SeismicAttributePanel(splitter_);
    attribute_panel_->setMaximumWidth(kPanelMaxWidth);
    splitter_->addWidget(attribute_panel_);
    view_panel_ = new SeismicViewPanel(splitter_);
    splitter_->addWidget(view_panel_);
    control_panel_ = new SeismicControlPanel(splitter_);
    control_panel_->setMaximumWidth(kPanelMaxWidth);
    splitter_->addWidget(control_panel_);
    splitter_->setStretchFactor(0, 0);
    splitter_->setStretchFactor(1, 1);
    splitter_->setStretchFactor(2, 0);
    splitter_->setSizes({280, 900, 280});
    outer->addWidget(splitter_, 1);

    // M6 float/dock is platform infra — the slice keeps panels docked and
    // wires the page-level signal graph verbatim (ledger note).
    connect(context_toolbar_, &SeismicContextToolbar::run_requested, this,
            &SeismicPredictionPage::on_run);
    connect(context_toolbar_, &SeismicContextToolbar::demo_requested, this,
            &SeismicPredictionPage::on_demo);
    connect(context_toolbar_, &SeismicContextToolbar::attribute_changed,
            this, &SeismicPredictionPage::on_attribute);
    connect(context_toolbar_,
            &SeismicContextToolbar::display_mode_changed, view_panel_,
            &SeismicViewPanel::set_display_mode);
    connect(context_toolbar_,
            &SeismicContextToolbar::display_mode_changed, control_panel_,
            &SeismicControlPanel::set_display_mode);
    connect(context_toolbar_, &SeismicContextToolbar::source_changed, this,
            [this](const QString& resource_id) {
                if (!resource_id.isEmpty()) {
                    select_seismic_resource(resource_id.toStdString());
                }
            });
    connect(control_panel_, &SeismicControlPanel::send_requested, this,
            &SeismicPredictionPage::send_to_mapping_requested);
    connect(control_panel_,
            &SeismicControlPanel::display_mode_changed, view_panel_,
            &SeismicViewPanel::set_display_mode);
    connect(attribute_panel_,
            &SeismicAttributePanel::attribute_changed, this,
            &SeismicPredictionPage::on_attribute);
    connect(control_panel_, &SeismicControlPanel::well_tie_toggled,
            view_panel_, &SeismicViewPanel::set_well_tie_enabled);
    connect(view_panel_, &SeismicViewPanel::view_ready, this,
            &SeismicPredictionPage::on_view_ready);
}

void SeismicPredictionPage::set_hooks(SeismicPredictionHooks hooks) {
    hooks_ = std::move(hooks);
}

SeismicContextToolbar* SeismicPredictionPage::context_toolbar() const {
    return context_toolbar_;
}
SeismicAttributePanel* SeismicPredictionPage::attribute_panel() const {
    return attribute_panel_;
}
SeismicViewPanel* SeismicPredictionPage::view_panel() const {
    return view_panel_;
}
SeismicControlPanel* SeismicPredictionPage::control_panel() const {
    return control_panel_;
}

void SeismicPredictionPage::set_project(const ProjectSlice* project) {
    if (project != project_) {
        ++session_token_;
        selected_resource_id_.reset();
        showing_selected_source_ = false;
    }
    project_ = project;
}

bool SeismicPredictionPage::shutdown_workers(int wait_ms) {
    const bool joined =
        !hooks_.shutdown || hooks_.shutdown(wait_ms);
    // A switch is aborted when a job cannot join — preserve the page and
    // session; the detached worker's results were already disconnected.
    if (!joined) {
        return false;
    }
    ++session_token_;
    inference_active_ = false;
    project_ = nullptr;
    view_panel_->shutdown();
    return true;
}

void SeismicPredictionPage::closeEvent(QCloseEvent* event) {
    // A page owns its inference thread; joining here covers direct widget
    // destruction as well as normal project-session shutdown.
    shutdown_workers();
    event->accept();
}

void SeismicPredictionPage::update_state(
    const std::vector<PredictionTaskSlice>& tasks,
    const ProjectSlice* project) {
    if (project != nullptr) {
        set_project(project);
    }
    tasks_ = tasks;
    sync_seismic_sources();
    const PredictionTaskSlice* task =
        showing_selected_source_ ? nullptr : current_task();
    const ResourceSlice* resource = selected_seismic_resource();
    if (showing_selected_source_ && resource != nullptr) {
        view_panel_->show_resource(*resource, project_);
    } else {
        view_panel_->update_state(task, project_);
    }
    control_panel_->update_state(task, view_panel_->volume_shape());
    sync_workbench_context(task);
    control_panel_->set_controls_enabled(view_panel_->is_view_ready());
}

const PredictionTaskSlice* SeismicPredictionPage::current_task() const {
    return active_prediction_task(tasks_);
}

void SeismicPredictionPage::sync_seismic_sources() {
    const std::string previous = selected_resource_id_.value_or("");
    const auto resources = segy_resources(project_);
    auto entries = source_combo_entries(resources, "未命名地震体", "");
    // Python parity: an empty list clears the combo (no placeholder row).
    context_toolbar_->set_source_entries(entries);
    const int selected =
        resolved_source_index(resources, previous);
    if (resources.empty() || selected < 0) {
        selected_resource_id_.reset();
        showing_selected_source_ = false;
    }
}

const ResourceSlice* SeismicPredictionPage::selected_seismic_resource()
    const {
    if (!selected_resource_id_.has_value() || project_ == nullptr) {
        return nullptr;
    }
    const auto resources = segy_resources(project_);
    return find_resource(resources, *selected_resource_id_);
}

std::optional<std::string>
SeismicPredictionPage::selected_seismic_resource_id() const {
    return selected_resource_id_;
}

bool SeismicPredictionPage::select_seismic_resource(
    const std::string& resource_id) {
    if (project_ == nullptr) {
        return false;
    }
    const auto resources = segy_resources(project_);
    const ResourceSlice* resource = find_resource(resources, resource_id);
    if (resource == nullptr) {
        return false;
    }
    selected_resource_id_ = resource->id;
    showing_selected_source_ = true;
    const bool loaded = view_panel_->show_resource(*resource, project_);
    control_panel_->update_state(nullptr, view_panel_->volume_shape());
    sync_workbench_context(nullptr);
    control_panel_->set_controls_enabled(view_panel_->is_view_ready());
    emit seismic_selection_changed(qs(*selected_resource_id_));
    return loaded;
}

void SeismicPredictionPage::on_attribute(const QString& label) {
    view_panel_->set_attribute_label(label);
    attribute_panel_->set_selected_attribute(label);
    sync_workbench_context(current_task());
}

void SeismicPredictionPage::on_view_ready(bool enabled) {
    control_panel_->set_controls_enabled(enabled);
    if (enabled) {
        const PredictionTaskSlice* task =
            showing_selected_source_ ? nullptr : current_task();
        sync_workbench_context(task);
    }
}

void SeismicPredictionPage::sync_workbench_context(
    const PredictionTaskSlice* task) {
    const QString attribute = view_panel_->attribute_label();
    const QString mode = view_panel_->display_mode();
    attribute_panel_->set_selected_attribute(attribute);
    control_panel_->set_attribute_label(attribute);
    // Python mirrors the control panel's horizon/shape/mock text into the
    // toolbar card; the same fields derive from the slice directly here.
    const std::string horizon =
        task != nullptr
            ? target_horizon_of(task->model_metadata, task->result_summary)
            : std::string();
    context_toolbar_->set_context(
        task, horizon, attribute.toStdString(), mode.toStdString(),
        view_panel_->volume_shape(),
        task != nullptr
            ? std::optional<std::string>(
                  seismic_output_nature(task->result_summary))
            : std::nullopt);
}

void SeismicPredictionPage::on_run() {
    // Production inference: resolve a production model, never auto-run
    // mock (spec P2 §3d — demo requires explicit demo mode).
    if (project_ == nullptr) {
        QMessageBox::warning(this, QStringLiteral("地震预测"),
                             QStringLiteral("未绑定工程，无法运行"));
        return;
    }
    if (!hooks_.catalog_connected || !hooks_.catalog_connected()) {
        QMessageBox::warning(this, QStringLiteral("地震预测"),
                             QStringLiteral("未连接数据目录，无法运行推断"));
        return;
    }
    const auto model_version_id =
        hooks_.production_model_id ? hooks_.production_model_id()
                                   : std::nullopt;
    if (!model_version_id.has_value()) {
        QMessageBox::warning(
            this, QStringLiteral("地震预测"),
            QStringLiteral(
                "未配置生产模型，无法运行科学预测。\n"
                "请先注册生产模型（ModelRegistry），"
                "或通过「运行演示预测」查看演示结果。"));
        return;
    }
    start_inference(*model_version_id, "seismic_facies", "地震相预测",
                    /*demo=*/false, selected_resource_id_);
}

void SeismicPredictionPage::on_demo() {
    // Explicit demo mode: run the registered DemoModelProvider.
    if (project_ == nullptr) {
        QMessageBox::warning(this, QStringLiteral("地震预测"),
                             QStringLiteral("未绑定工程，无法运行"));
        return;
    }
    if (!hooks_.catalog_connected || !hooks_.catalog_connected()) {
        QMessageBox::warning(this, QStringLiteral("地震预测"),
                             QStringLiteral("未连接数据目录，无法运行推断"));
        return;
    }
    const auto demo_version_id =
        hooks_.demo_model_id ? hooks_.demo_model_id() : std::nullopt;
    if (!demo_version_id.has_value()) {
        QMessageBox::warning(this, QStringLiteral("地震预测"),
                             QStringLiteral("演示模型未注册"));
        return;
    }
    start_inference(*demo_version_id, "seismic_facies",
                    "地震相预测(Demo)", /*demo=*/true,
                    selected_resource_id_);
}

void SeismicPredictionPage::start_inference(
    const std::string& model_version_id, const std::string& workflow,
    const std::string& name_prefix, bool demo,
    const std::optional<std::string>& resource_id) {
    if (hooks_.is_running && hooks_.is_running()) {
        return;
    }
    std::vector<std::string> input_ids;
    if (demo && !resource_id.has_value()) {
        input_ids.clear();
    } else if (hooks_.resolve_inputs) {
        const auto error = hooks_.resolve_inputs(model_version_id,
                                                 resource_id, input_ids);
        if (error.has_value()) {
            QMessageBox::warning(
                this, QStringLiteral("地震预测"),
                QStringLiteral("输入不满足模型契约: %1")
                    .arg(qs(*error)));
            return;
        }
    }
    domain::Json parameters = domain::Json::object();
    parameters["seed"] = static_cast<long long>(tasks_.size());
    parameters["workflow"] = workflow;
    parameters["name_prefix"] = name_prefix;
    parameters["demo"] = demo;
    parameters["seismic_resource_ids"] =
        resource_id.has_value()
            ? domain::Json::array({*resource_id})
            : domain::Json::array();
    inference_active_ = true;
    context_toolbar_->set_inferring(true);
    if (hooks_.start_run) {
        hooks_.start_run(model_version_id, input_ids, parameters);
    }
}

void SeismicPredictionPage::on_inference_completed(
    const domain::Json& payload) {
    // Session-token parity: a stale worker (project switched / shutdown)
    // cannot publish into the rebound page.
    if (!inference_active_) {
        return;
    }
    inference_active_ = false;
    context_toolbar_->set_inferring(false);
    const Json& run = json_field(payload, "run");
    const std::string run_status =
        run.is_object() ? json_str(run, "status", "") : "";
    if (!run.is_object() || run_status == "failed") {
        std::string error = "未知错误";
        if (run.is_object()) {
            error = json_str(json_object(run, "parameters"), "error",
                             "未知错误");
        }
        // Async completion: in-page status instead of a modal dialog.
        context_toolbar_->set_status(
            QStringLiteral("推断失败: %1").arg(qs(error)));
        return;
    }
    // Terminal cancel: honest status — never a fabricated completion and
    // never a failure scare (the partial tiles stay resume-eligible).
    if (run_status == "cancelled") {
        context_toolbar_->set_status(QStringLiteral("推断已取消"));
        return;
    }
    const Json& result = json_field(payload, "result");
    if (!result.is_object() || result.empty()) {
        const std::string error =
            run.is_object()
                ? json_str(json_object(run, "parameters"), "error",
                           "预测完成但未返回可用结果")
                : "预测完成但未返回可用结果";
        context_toolbar_->set_status(
            QStringLiteral("推断失败: %1").arg(qs(error)));
        return;
    }
    if (hooks_.materialize_task) {
        tasks_.push_back(hooks_.materialize_task(run, result));
    }
    showing_selected_source_ = false;
    update_state(tasks_, project_);
    emit prediction_updated();
}

void SeismicPredictionPage::on_inference_failed(const std::string& text) {
    if (!inference_active_) {
        return;
    }
    inference_active_ = false;
    context_toolbar_->set_inferring(false);
    context_toolbar_->set_status(
        QStringLiteral("推断失败: %1").arg(qs(text)));
}

}  // namespace pwb::ui_wellseis::qt
